File: Python/Reverse-Linked-List-II.py
#!/usr/bin/env python

# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    # @param head, a ListNode
    # @param m, an integer
    # @param n, an integer
    # @return a ListNode
    # m = 2, n = 4
    def reverseBetween(self, head, m, n):
        if m == n:
            return head
        start = head
        end = head
        prenode = None
        for i in range(m - 1):
            prenode = start
            start = start.next
            end = end.next
        for i in range(n - m):
            end = end.next
        new_end = end.next
        while start != end:
            node = start.next
            start.next = new_end
            new_end = start
            start = node
        if prenode == None:
            prenode = start
            start.next = new_end
            return prenode
        else:
            start.next = new_end
            prenode.next = start
            return head

File: C++/Reverse-Linked-List-II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverse(ListNode* head) {
        ListNode* res = NULL;
        while (head != NULL) {
            ListNode* temp = head->next;
            head->next = res;
            res = head;
            head = temp;
        }
        return res;
    }
    ListNode* reverseBetween(ListNode* head, int m, int n) {
        if (m == n) return head;
        ListNode *pre = new ListNode(0), *cur = pre;
        pre->next = head;
        for (int i = 1; i < m; i ++) {
            cur = cur->next;
        }
        ListNode* pre_start = cur;
        for (int i = 0; i < n - m; i ++) {
            cur = cur->next;
        }
        ListNode *pre_end = cur, *start = pre_start->next;
        ListNode* next_end = cur->next->next;
        cur->next->next = NULL;
        ListNode *new_start = reverse(start);
        pre_start->next = new_start;
        cur = pre;
        while (cur->next != NULL) {
            cur = cur->next;
        }
        cur->next = next_end;
        return pre->next;
    }
};